File: alpine-linux.py
import json
import os
from common import Storage

storage = Storage('alpine-linux')


def main():
    remote_site = 'https://dl-cdn.alpinelinux.org/alpine'
    versions = [
        'v3.0', 'v3.1', 'v3.2', 'v3.3', 'v3.4', 'v3.5', 'v3.6', 'v3.7', 'v3.8', 'v3.9', 'v3.10',
        'v3.11', 'v3.12', 'v3.13', 'v3.14', 'v3.15', 'v3.16', 'v3.17', 'v3.18'
    ]
    for version in versions:
        items = json.loads(
            os.popen('rclone lsjson -R --include=\'/**.iso\' --http-url {}/{}/releases/ :http:'.format(remote_site, version)).read())
        for v in items:
            if v['IsDir']:
                continue
            storage.write('{}/{}'.format(version, v['Path']),
                          '{}/{}/releases/{}'.format(remote_site, version, v['Path']))


if __name__ == '__main__':
    main()

File: app/data/task/alpinelinux.go
package task

import (
	"fmt"

	"go.uber.org/zap"

	"github.com/ismdeep/mirror-data/app/data/conf"
	"github.com/ismdeep/mirror-data/app/data/global"
	"github.com/ismdeep/mirror-data/app/data/internal/rclone"
	"github.com/ismdeep/mirror-data/app/data/internal/store"
	util2 "github.com/ismdeep/mirror-data/app/data/internal/util"
	"github.com/ismdeep/mirror-data/pkg/log"
)

type AlpineLinux struct {
}

func (receiver *AlpineLinux) Run() {
	storage := store.New("alpine-linux", conf.Config.StorageCoroutineSize)

	remoteSite := "https://dl-cdn.alpinelinux.org/alpine"

	versionsResp, err := rclone.JSON("lsjson", "--http-url", remoteSite, ":http:")
	if err != nil {
		global.Errors <- err
	}

	versionChan := make(chan string, 1024)
	go func() {
		defer func() {
			close(versionChan)
		}()

		for _, vResp := range versionsResp {
			if util2.IsAlpineVersion(vResp.Path) {
				versionChan <- vResp.Path
			}
		}
	}()

	err = util2.CoroutineRun(conf.Config.CoroutineSize, func() error {
		for version := range versionChan {
			items, err := rclone.JSON("lsjson", "-R", "--http-url", fmt.Sprintf("%v/%v/releases/", remoteSite, version), ":http:")
			if err != nil {
				log.WithName("alpine-linux").Error("failed on lsjson", zap.Error(err))
				return err
			}
			for _, v := range items {
				if v.IsDir {
					continue
				}

				if util2.StringEndWith(v.Path, ".iso") {
					storage.Add(
						fmt.Sprintf("%v/%v", version, v.Path),
						fmt.Sprintf("%v/%v/releases/%v", remoteSite, version, v.Path))
				}
			}
		}
		return nil
	})
	if err != nil {
		global.Errors <- err
		return
	}

	storage.CloseAndWait()
}

func (receiver *AlpineLinux) GetBucketName() string {
	return "alpine-linux"
}
